File: src/objective-c/examples/Sample/Sample/ViewController.m
/*
 *
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "ViewController.h"

#import <GRPCClient/GRPCCall.h>
#import <ProtoRPC/ProtoMethod.h>
#import <RemoteTest/Messages.pbobjc.h>
#import <RemoteTest/Test.pbrpc.h>
#import <RxLibrary/GRXWriter+Immediate.h>
#import <RxLibrary/GRXWriteable.h>

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  NSString * const kRemoteHost = @"grpc-test.sandbox.googleapis.com";

  RMTSimpleRequest *request = [[RMTSimpleRequest alloc] init];
  request.responseSize = 10;
  request.fillUsername = YES;
  request.fillOauthScope = YES;

  // Example gRPC call using a generated proto client library:

  RMTTestService *service = [[RMTTestService alloc] initWithHost:kRemoteHost];
  [service unaryCallWithRequest:request handler:^(RMTSimpleResponse *response, NSError *error) {
    if (response) {
      NSLog(@"Finished successfully with response:\n%@", response);
    } else if (error) {
      NSLog(@"Finished with error: %@", error);
    }
  }];


  // Same example call using the generic gRPC client library:

  GRPCProtoMethod *method = [[GRPCProtoMethod alloc] initWithPackage:@"grpc.testing"
                                                             service:@"TestService"
                                                              method:@"UnaryCall"];

  GRXWriter *requestsWriter = [GRXWriter writerWithValue:[request data]];

  GRPCCall *call = [[GRPCCall alloc] initWithHost:kRemoteHost
                                             path:method.HTTPPath
                                   requestsWriter:requestsWriter];

  id<GRXWriteable> responsesWriteable = [[GRXWriteable alloc] initWithValueHandler:^(NSData *value) {
    RMTSimpleResponse *response = [RMTSimpleResponse parseFromData:value error:NULL];
    NSLog(@"Received response:\n%@", response);
  } completionHandler:^(NSError *errorOrNil) {
    if (errorOrNil) {
      NSLog(@"Finished with error: %@", errorOrNil);
    } else {
      NSLog(@"Finished successfully.");
    }
  }];

  [call startWithWriteable:responsesWriteable];
}

@end

File: src/objective-c/examples/SwiftSample/ViewController.swift
/*
 *
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import UIKit

import RemoteTest

class ViewController: UIViewController {

  override func viewDidLoad() {
    super.viewDidLoad()

    let RemoteHost = "grpc-test.sandbox.googleapis.com"

    let request = RMTSimpleRequest()
    request.responseSize = 10
    request.fillUsername = true
    request.fillOauthScope = true


    // Example gRPC call using a generated proto client library:

    let service = RMTTestService(host: RemoteHost)
    service.unaryCall(with: request) { response, error in
      if let response = response {
        NSLog("1. Finished successfully with response:\n\(response)")
      } else {
        NSLog("1. Finished with error: \(error!)")
      }
    }


    // Same but manipulating headers:

    var RPC : GRPCProtoCall! // Needed to convince Swift to capture by reference (__block)
    RPC = service.rpcToUnaryCall(with: request) { response, error in
      if let response = response {
        NSLog("2. Finished successfully with response:\n\(response)")
      } else {
        NSLog("2. Finished with error: \(error!)")
      }
      NSLog("2. Response headers: \(String(describing: RPC.responseHeaders))")
      NSLog("2. Response trailers: \(String(describing: RPC.responseTrailers))")
    }

    // TODO(jcanizales): Revert to using subscript syntax once XCode 8 is released.
    RPC.requestHeaders["My-Header"] = "My value"

    RPC.start()


    // Same example call using the generic gRPC client library:

    let method = GRPCProtoMethod(package: "grpc.testing", service: "TestService", method: "UnaryCall")!

    let requestsWriter = GRXWriter(value: request.data())!

    let call = GRPCCall(host: RemoteHost, path: method.httpPath, requestsWriter: requestsWriter)!

    call.requestHeaders["My-Header"] = "My value"

    call.start(with: GRXWriteable { response, error in
      if let response = response as? Data {
        NSLog("3. Received response:\n\(try! RMTSimpleResponse(data: response))")
      } else {
        NSLog("3. Finished with error: \(error!)")
      }
      NSLog("3. Response headers: \(String(describing: call.responseHeaders))")
      NSLog("3. Response trailers: \(String(describing: call.responseTrailers))")
    })
  }
}
